IRGen and runtime support for key path resilience.

jckarter opened the pull request:
IRGen support for emitting computed key path components with references to external property descriptors, and matching runtime support for instantiating references to external property descriptors into key path objects.

Includes a fix for rdar://problem/39196876.

jckarter committed:
Remove stale -enable-sil-ownership flag from test

jckarter committed:
Adjust the key path component tags so that the trivial property descriptor can be zero.

…and thus more likely to be folded with another zero constant in the module and/or put in a zero-initialized page.

jckarter committed:
SILGen: Fix key paths that reference internal private(set) decls from other files.

The setter needs to be given hidden linkage so that other files can still form key paths to it.

jckarter committed:
IRGen: Implement codegen for external key path components with local attempts.

Include the information necessary to instantiate the external component (if any) before the local attempt at forming a component.

jckarter committed:
IRGen: Fix null pointer deref in nongeneric external property ref.

jckarter committed:
KeyPaths: Stub out support for instantiating external components with trivial descriptors.

jckarter committed:
TBDGen: Include property descriptor in TBD output.

jckarter committed:
KeyPaths: Instantiate external property descriptors for stored properties.

jckarter committed:
SILGen: Fix visibility handling of external private setters in key paths.

Don't try to directly reference them if they can't be linked to from here.

jckarter committed:
KeyPaths: Handle computed property descriptor instantiation w/o indexes or generics.

jckarter committed:
SIL: Tighten verification rule for property descriptors.

A property descriptor shouldn't refer to yet another external declaration.

slavapestov commented on test/SILGen/keypaths.swift:
```diff
@@ -1,5 +1,4 @@
-
-// RUN: %target-swift-emit-silgen -module-name keypaths -enable-sil-ownership %s | %FileCheck %s
+// RUN: %target-swift-emit-silgen -module-name keypaths %s | %FileCheck %s
```
Is this always on by default now?

jckarter replied:
I believe so, yeah. cc @gottesmm

jckarter committed:
KeyPaths: Rationalize fragility attributes.

jckarter committed:
Remove test/stdlib/KeyPathImplementation test.

This was useful for bringup before compiler support for building key path patterns was implemented, but now it just hardcodes a bunch of implementation details that expose some internal details as ABI and interfere with refactorings. Remove it now that the same functionality is tested at a higher level.

jckarter committed:
KeyPaths: Centralize some flag testing logic.

jckarter committed:
KeyPaths: Support instantiating generic property descriptors from concrete contexts.

This is a bit easier than the fully general case where both the external descriptor and local pattern have captured arguments (because of generics or subscript indices) since we don't have to combine the two argument files in one component.

jckarter committed:
KeyPaths: Support instantiating property descriptors with captured arguments.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test
